BUG: Fix reindexing with multi-indexed DataFrames

ChrisRobo opened the pull request:
- [x] closes https://github.com/pandas-dev/pandas/issues/29896
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Addresses an issue which appears to have existed since 0.23.0 where bugs in the `get_indexer()` method for the `MultiIndex` class cause incorrect reindexing behavior for multi-indexed DataFrames.

motivating, example, from (the issue)
```python
>>> 
>>> df = pd.DataFrame({
...     'a': [0, 0, 0, 0],
...     'b': [0, 2, 3, 4],
...     'c': ['A', 'B', 'C', 'D']
... }).set_index(['a', 'b'])
>>> 
>>> df
     c
a b   
0 0  A
  2  B
  3  C
  4  D
>>> df.index
MultiIndex([(0, 0),
            (0, 2),
            (0, 3),
            (0, 4)],
           names=['a', 'b'])
>>> mi_2 = pd.MultiIndex.from_product([[0], [-1, 0, 1, 3, 4, 5]])
>>> mi_2
MultiIndex([(0, -1),
            (0,  0),
            (0,  1),
            (0,  3),
            (0,  4),
            (0,  5)],
           )
```

as expected, without a `method` value:
```python
>>> df.reindex(mi_2)
        c
0 -1  NaN
   0    A
   1  NaN
   3    C
   4    D
   5  NaN
```

using `method="backfill"`, it is:
```python
>>> 
>>> df.reindex(mi_2, method="backfill")
      c
0 -1  A
   0  A
   1  D
   3  A
   4  A
   5  C
>>>
```
but should (IMHO) be:
```python
>>> df.reindex(mi_2, method="backfill")
        c
0 -1    A
   0    A
   1    B
   3    C
   4    D
   5  NaN
>>>
```

similarly, using `method="pad"`, it is:
```python
>>> df.reindex(mi_2, method="pad")
        c
0 -1  NaN
   0  NaN
   1    D
   3  NaN
   4    A
   5    C
```
but should (IMHO) be:
```python
>>> df.reindex(mi_2, method="pad")
        c
0 -1  NaN
   0    A
   1    A
   3    C
   4    D
   5    D
```

ChrisRobo committed:
* add missing method kwargs in MultiIndex get_indexer() impl

* add tests for functions covered by bugfix

ChrisRobo committed:
handle multiple trickier cases with indexing

* handle tuple-ordering case, i.e. the highest-level (highest being closest
to 0, i.e. the leftmost) fill up or down, for backfill/pad, respectively,
should override values at all lower levels to their lowest/highest
values, respectively

* stub out handling of "carrying" situation with too-large values at
some level which may require values at higher levels being bumped

ChrisRobo committed:
add small bugfixes and cleanup

* fix bug with incorrect handling of legit settling at max val at some level

* use np arrays for checking test outputs

* dont overflow from a val when we were backfilled up to it

* remove debugging-only code, no change to logic

ChrisRobo committed:
fix more bugs and do more cleanup

* clean up comments

* small cleanup and fix for tuple-order-handling fix

* fix ordering of extract_level_codes() cleanup

* add test for case with fill at multiple levels

* rm duplicative test

* move carrying code into its own function

ChrisRobo committed:
update tests to reflect changes since 0.24.2

ChrisRobo committed:
handle tuple-ordering correctly across several levels when carrying

ChrisRobo committed:
rm testing print statements and fix rebase error

ChrisRobo committed:
address flake8 style complaints

ChrisRobo committed:
apply output of

pep8speaks commented:
Hello @ChrisRobo! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/tests/frame/indexing/test_indexing.py`](https://github.com/pandas-dev/pandas/blob/9a1d147efe86bb2a7259765054abbb5eedebd298/pandas/tests/frame/indexing/test_indexing.py):

> [Line 1591:68](https://github.com/pandas-dev/pandas/blob/9a1d147efe86bb2a7259765054abbb5eedebd298/pandas/tests/frame/indexing/test_indexing.py#L1591): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1597:75](https://github.com/pandas-dev/pandas/blob/9a1d147efe86bb2a7259765054abbb5eedebd298/pandas/tests/frame/indexing/test_indexing.py#L1597): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1600:61](https://github.com/pandas-dev/pandas/blob/9a1d147efe86bb2a7259765054abbb5eedebd298/pandas/tests/frame/indexing/test_indexing.py#L1600): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1603:61](https://github.com/pandas-dev/pandas/blob/9a1d147efe86bb2a7259765054abbb5eedebd298/pandas/tests/frame/indexing/test_indexing.py#L1603): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

* In the file [`pandas/tests/indexes/multi/test_indexing.py`](https://github.com/pandas-dev/pandas/blob/9a1d147efe86bb2a7259765054abbb5eedebd298/pandas/tests/indexes/multi/test_indexing.py):

> [Line 279:64](https://github.com/pandas-dev/pandas/blob/9a1d147efe86bb2a7259765054abbb5eedebd298/pandas/tests/indexes/multi/test_indexing.py#L279): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 283:53](https://github.com/pandas-dev/pandas/blob/9a1d147efe86bb2a7259765054abbb5eedebd298/pandas/tests/indexes/multi/test_indexing.py#L283): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 287:53](https://github.com/pandas-dev/pandas/blob/9a1d147efe86bb2a7259765054abbb5eedebd298/pandas/tests/indexes/multi/test_indexing.py#L287): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

ChrisRobo commented:
> Hello @ChrisRobo! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:
> 
> * In the file [`pandas/tests/frame/indexing/test_indexing.py`](https://github.com/pandas-dev/pandas/blob/26c8ac3e150a8559ee5f372b0b66db44c9d2f3c2/pandas/tests/frame/indexing/test_indexing.py):
> 
> > [Line 1591:68](https://github.com/pandas-dev/pandas/blob/26c8ac3e150a8559ee5f372b0b66db44c9d2f3c2/pandas/tests/frame/indexing/test_indexing.py#L1591): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> > [Line 1597:75](https://github.com/pandas-dev/pandas/blob/26c8ac3e150a8559ee5f372b0b66db44c9d2f3c2/pandas/tests/frame/indexing/test_indexing.py#L1597): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> > [Line 1600:61](https://github.com/pandas-dev/pandas/blob/26c8ac3e150a8559ee5f372b0b66db44c9d2f3c2/pandas/tests/frame/indexing/test_indexing.py#L1600): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> > [Line 1603:61](https://github.com/pandas-dev/pandas/blob/26c8ac3e150a8559ee5f372b0b66db44c9d2f3c2/pandas/tests/frame/indexing/test_indexing.py#L1603): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> 
> * In the file [`pandas/tests/indexes/multi/test_indexing.py`](https://github.com/pandas-dev/pandas/blob/26c8ac3e150a8559ee5f372b0b66db44c9d2f3c2/pandas/tests/indexes/multi/test_indexing.py):
> 
> > [Line 279:64](https://github.com/pandas-dev/pandas/blob/26c8ac3e150a8559ee5f372b0b66db44c9d2f3c2/pandas/tests/indexes/multi/test_indexing.py#L279): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> > [Line 283:53](https://github.com/pandas-dev/pandas/blob/26c8ac3e150a8559ee5f372b0b66db44c9d2f3c2/pandas/tests/indexes/multi/test_indexing.py#L283): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> > [Line 287:53](https://github.com/pandas-dev/pandas/blob/26c8ac3e150a8559ee5f372b0b66db44c9d2f3c2/pandas/tests/indexes/multi/test_indexing.py#L287): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> 
> ##### Comment last updated at 2020-01-06 23:44:11 UTC

the formatting decisions on these lines are auto-generated output of `black pandas`.  I defer to the reviewers here re: the preferred style decisions

ChrisRobo committed:
fix test import names and ordering

ChrisRobo committed:
move whatsnew addition from Indexing to MultiIndex section

ChrisRobo commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -183,6 +183,155 @@ def test_get_indexer():
         idx1.get_indexer(idx2)
 
 
+def test_get_indexer_and_fill():
```
in all of these test cases, they illustrate the behavior up until 0.23.0.  The assumption of this PR is that, in the absence of documentation about these changes to MultiIndex behavior in 0.23.0 (and that this PR does not break any existing tests), that those changes are bugs

ChrisRobo committed:
address linting issues

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -653,6 +653,86 @@ cdef class BaseMultiIndexCodesEngine:
+    def _do_backfill_carrying(self, object level_codes,
+                              object level_codes_no_fill):
+        """
+        given a 2d list of level_codes, i.e. integers representing the index
```
start with a 1-line summary

ChrisRobo replied:
added, thanks

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -653,6 +653,86 @@ cdef class BaseMultiIndexCodesEngine:
+
+        Parameters
+        ----------
+        level_codes : 2D array, M x N (where self has N levels)
```
can the inputs be typed as `ndarray[ndim=2]`?

ChrisRobo replied:
good idea, updated

jbrockmendel commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -183,6 +183,155 @@ def test_get_indexer():
 
 
+def test_get_indexer_and_fill():
+    """ test getting an indexer for another index using the backfill method """
```
comment instead of docstring, add a reference to the original GH sissue

ChrisRobo replied:
done, thanks for the tip

jbrockmendel commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -183,6 +183,155 @@ def test_get_indexer():
+    for method in ("ffill", "pad"):
+        tm.assert_almost_equal(
+            mult_idx_1.get_indexer(mult_idx_2, method=method), np.array([0, 2, 3])
+        )
```
separate lines

result = mult_idx...
expected = np.array...
tm.assert_almost_equal(result, expected)

ChrisRobo replied:
done (here and elsewhere), good idea; thanks

jreback reviewed (changes requested):
i don’t think the reindeximg result is valid here

pls show your original example at the top of the PR
include the case of reindex without pad/backfill

show the same example with an Index (factorize the tuples); 

convince us why these should be different as that is what you are proposing (i think)

ChrisRobo committed:
use intermediary variables in tests and add issue link in comments

ChrisRobo committed:
add one-line summary for _do_backfill_carrying() method

ChrisRobo committed:
use comments instead of docstrings with issue link in tests

ChrisRobo commented:
> i don’t think the reindeximg result is valid here
> 
> pls show your original example at the top of the PR
> include the case of reindex without pad/backfill
> 
> show the same example with an Index (factorize the tuples);
> 
> convince us why these should be different as that is what you are proposing (i think)

thanks for the quick feedback here!

I included the original example from the issue (with both what I saw and what I expected to see) and also added the case of no filling (not included in the issue because I don't view that as a bug).

Apologies for the ignorance (this is my first PR here), but what specifically are you looking for with tuple factorization?  I'm having trouble constructing an index of tuples which is not itself a MultiIndex.

My argument for why this is a bug is that backfilling/padding should respect python tuple ordering properties (which it appears to have done until version 0.23.0), e.g. `(0, 0) < (0, 1) < (0, 2)`, and so, using the case of the example
- when backfilling, the row with index value (0, 1) should have `B` (not `D`) as its `c` value, since this should be backfilled to the row with index value (0, 2) in `df` (whose `c` value is `B`)
- when padding, the row with index value (0, 1) should have `A` (not `D`) as its `c` value, since this should be padded to the row with index value (0, 0) in `df` (whose `c` value is `A`)
- moreover, in both cases, the rows with index values (0, 3) and (0, 4) should have `c` values `C` and `D` respectively (they are `A` and `A` with backfilling and `NaN` and `A` with padding), since (0, 3) and (0, 4) are both in `df.index` to begin with

Happy to add more examples or any kind of additional explanation or validation that you'd like to help make this clearer.  Thanks again for the responsiveness

ChrisRobo committed:
rm unnecessary blank line

ChrisRobo commented:
here's a more concise version of the example from the GH issue which still illustrates the problem, I think

setup:
```python
>>> df = pd.DataFrame({'a': [0, 0], 'b': [0, 2], 'c': ['A', 'B']}).set_index(['a', 'b'])
>>> df
     c
a b   
0 0  A
  2  B
>>> 
>>> mi_2 = pd.MultiIndex.from_product([[0], [1]])
>>> mi_2
MultiIndex([(0, 1)],
           )
```

in both cases, without a `method` value:
```python
>>> df.reindex(mi_2)
       c
0 1  NaN
```

filling with `backfill`, as is (which I argue is a bug):
```python
>>> df.reindex(mi_2, method='bfill')
     c
0 1  A
```
I believe it should be:
```python
>>> df.reindex(mi_2, method='bfill')
     c
0 1  B
```
similarly, as is, with `pad` (I also believe this is a bug):
```python
>>> df.reindex(mi_2, method='pad')
       c
0 1  NaN
```
whereas I believe it should be
```
>>> df.reindex(mi_2, method='pad')
     c
0 1  A
```

hopefully this is easier to read

ChrisRobo committed:
use indexer dtypes in tests for compatibility with windows

ChrisRobo committed:
one more such case

ChrisRobo committed:
cleanup for same such cases

ChrisRobo committed:
run black pandas again, update formatting of a few lines

jreback commented:
@ChrisRobo can you update your example to show a few values, instead of a single one. some should *have* values even in the reindex and some should not so as to see the effect of the reindex.

ChrisRobo commented:
> @ChrisRobo can you update your example to show a few values, instead of a single one. some should _have_ values even in the reindex and some should not so as to see the effect of the reindex.

sure (making a new comment in the interest of preserving the history of this thread)

here's an example which I think is a bit more comprehensive.

setup:
```python
>>> df = pd.DataFrame({
...     'a': [1, 1, 1, 4, 4, 4],
...     'b': [1, 5, 9, 1, 5, 9],
...     'c': ['A', 'B', 'C', 'D', 'E', 'F'],
... })
>>> df = df.set_index(['a', 'b'])
>>> df
     c
a b   
1 1  A
  5  B
  9  C
4 1  D
  5  E
  9  F
>>> 
>>> mi_2 = pd.MultiIndex.from_tuples([
...     (1, 0),
...     (1, 5),
...     (1, 7),
...     (3, 2),
...     (4, 9),
...     (4, 11),
...     (5, 2),
... ])
>>> mi_2
MultiIndex([(1,  0),
            (1,  5),
            (1,  7),
            (3,  2),
            (4,  9),
            (4, 11),
            (5,  2)],
           )
>>> 
```
current behavior w/o a filling method (no proposed change):
```python
>>> df.reindex(mi_2)
        c
1 0   NaN
  5     B
  7   NaN
3 2   NaN
4 9     F
  11  NaN
5 2   NaN
```
current behavior w/`backfill` and `pad`:
```python
>>> df.reindex(mi_2, method='backfill')
      c
1 0   A
  5   F
  7   A
3 2   A
4 9   A
  11  D
5 2   B
>>> 
>>> 
>>> df.reindex(mi_2, method='pad')
        c
1 0   NaN
  5     F
  7   NaN
3 2   NaN
4 9   NaN
  11    C
5 2     B
```
whereas I think this is correct:
```python
>>> df.reindex(mi_2, method='backfill')
        c
1 0     A
  5     B
  7     C
3 2     D
4 9     F
  11  NaN
5 2   NaN
>>> 
>>> df.reindex(mi_2, method='pad')
        c
1 0   NaN
  5     B
  7     B
3 2     C
4 9     F
  11    F
```

this covers the case of values which are too large/too small (for backfilling/padding, respectively), cases which exist in the original and new index (i.e. (1, 5) and (4, 11)), values in the new index to which nothing is filled (i.e. (4, 5, E)), and backfilling/padding cases (i.e. those which are `NaN` without a `method` value but are not afterward).

Let me know if there's a case which this is missing but which you think it should cover!

ChrisRobo committed:
move whatsnew entry from 1.0.0 -> 1.0.1

ChrisRobo committed:
actually, mv it to 1.1.0 since that appears to be next release

ChrisRobo committed:
sort imports in indexing tests with isort

ChrisRobo committed:
revert unrelated change to not-used 1.0.1 release notes

ChrisRobo commented:
@jreback  @jbrockmendel when you get a chance, would one or both of you mind taking another look at this?

I'm happy to add more examples or test coverage if you'd like

jreback commented:
will look. thought this was going to conflict with a recently merged refactor of MI indexing. pls merge master anyhow.

ChrisRobo commented:
> will look. thought this was going to conflict with a recently merged refactor of MI indexing. pls merge master anyhow.

thanks! :).  just rebased on top of master

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -649,7 +649,89 @@ cdef class BaseMultiIndexCodesEngine:
+
+        Parameters
+        ----------
+        level_codes : ndarray[ndim=2]
```
do we have a dtype here?  or maybe its int-like but we dont know which intlike?

ChrisRobo replied:
good point -- this method should work with any int-like type but its only use is with an array with `dtype='uint64'` (see [here](https://github.com/pandas-dev/pandas/pull/30766/files#diff-495f440aad68721ad5ca1f35087fc450R783), specifically).

what's the preferred way to represent a 2d np array of int-like values in a docstring here?

jbrockmendel replied:
`ndarray[uint64_t, ndim=2]`, which you might be able to add as a type in the signature

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -649,7 +649,89 @@ cdef class BaseMultiIndexCodesEngine:
+            per-level backfilled codes of a target index. a 0 represents NaN,
+            and the remainder of the values should be from 1, ..., L_i, where
+            L_i is the length of `self.levels[i]`
+        level_codes_no_fill: 2D-list-like, M x N
```
since this is very-internal, can we be any more specific/strict than list-like?

ChrisRobo replied:
sure -- it's also a 2d np.array of ints, so will use the same formatting as preferred above to describe this

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -172,6 +172,7 @@ MultiIndex
                           index=[["a", "a", "b", "b"], [1, 2, 1, 2]])
         # Rows are now ordered as the requested keys
         df.loc[(['b', 'a'], [2, 1]), :]
+- Bug in :meth:`MultiIndex.get_indexer` incorrectly handling use of pad and backfill options (:issue:`29896`)
```
can you make a separate sub section showing the before and the new results; yes its a bug fix but the effect is non-obvious. you can use the OP example.

ChrisRobo replied:
done, let me know if the formatting is OK by you!

jreback commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -183,6 +183,170 @@ def test_get_indexer():
+    # https://github.com/pandas-dev/pandas/issues/29896
+    # test getting an indexer for another index with different methods
+    mult_idx_1 = MultiIndex.from_product([[0], [0, 2, 3, 4]])
+    mult_idx_2 = MultiIndex.from_product([[0], [1, 3, 4]])
```
example to include at least 2 0-levels, e.g. the backfill / pad operations should not cross the boundaries (doesn't have to be on every one, just some that are relevant)

ChrisRobo replied:
done, thanks for the suggestion!

jreback commented on pandas/tests/frame/indexing/test_indexing.py:
```diff
@@ -1590,6 +1590,42 @@ def test_reindex_methods(self, method, expected_values):
         tm.assert_frame_equal(expected, actual)
 
+    def test_reindex_with_multi_index(self):
+        # https://github.com/pandas-dev/pandas/issues/29896
```
same comment on the tests below

ChrisRobo replied:
done

jreback reviewed (changes requested):
this is a *very* complicated implementation which is adding quite a bit of code. It seems to me that you can simply re-order the codes to avoid the carrying operation entirely, no? this is similar to recoding the values due to a new factorization.

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -649,7 +649,89 @@ cdef class BaseMultiIndexCodesEngine:
+            performed, where values which are too large are represented by
+            rows consisting of all 0s
+        """
+        for i, row in enumerate(level_codes):
```
can you add types for i, j, really whatever else you can.  it makes a difference for performance

ChrisRobo replied:
done (in re-written implementation)

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -658,19 +740,54 @@ cdef class BaseMultiIndexCodesEngine:
+            # necessary to respect tuple ordering.  intuition is that bumping
+            # the value at level i should make the values at levels i+1, ..., n
+            # as small as possible, and vice versa
+            for i, row in enumerate(level_codes):
```
same, please add types

ChrisRobo replied:
done (in re-written implementation)

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -658,19 +740,54 @@ cdef class BaseMultiIndexCodesEngine:
+        level_codes = np.array([
+            lev.get_indexer(codes, method=method) for lev, codes
+            in zip(self.levels, zip(*target))
+        ], dtype='uint64').T + 1
```
can any of this go directly in MultiIndex and not in the cython code?  the for-loops clearly belong here, but to the extent we can keep Index objects out of this, we should

ChrisRobo replied:
the updated version does not add any new uses of Index objects in the cython code

ChrisRobo committed:
add example code in whatsnew

ChrisRobo committed:
re-write backend for reindexing MultiIndexes with a filling method

* use approach similar to creating a new factorization of a merged array
of the MultiIndex values

* add explanatory comments

* separate implementation of MultiIndex.get_indexer() with and without
filling; former is unchanged

* use np C types strictly in cython code and add type annotations where
possible

ChrisRobo committed:
* add docstings, fix comments

* simplify code and make get_indexer_and_fill method static

ChrisRobo committed:
robustify tests and add additional comments

ChrisRobo committed:
include before/after examples

ChrisRobo commented:
@jreback  @jbrockmendel thanks for the suggestions, sorry for the late reply here.

I basically re-wrote this using what (I think is) a much simpler implementation, i.e. using something like an updated factorization of the MultiIndexes' tuples and passing that to the relevant filling method.  I also added types wherever I could in cython code and made additions to the tests as requested.

ChrisRobo commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,112 @@ cdef class BaseMultiIndexCodesEngine:
+        # merger of the index values, where `new_codes` and `new_target_codes`
+        # are the subset of the factors which appear in `values` and `target`,
+        # respectively
+        i, j, next_code = 0, 0, 0
```
example of how this code works:

if `values` is
```
[
    (1, 1),
    (1, 3),
    (2, 1),
    (2, 3),
]
```

and `target_values` (i.e. `target.values`, sorted) is:
```
[
   (0, 1),
   (1, 0),
   (1, 3),
   (2, 2),
   (2, 5),
]
```

then, if you merged these into a sorted list of (unique) values, you would get (including index values here, and an X indicates membership):
```
          values  target_values

0: (0, 1)               X
1: (1, 0)               X
2: (1, 1)    X
3: (1, 3)    X          X
4: (2, 1)    X
5: (2, 2)               X
6: (2, 3)    X
7: (2, 5)               X
```
and so we can think of this as a factorization of the merged and sorted list of tuples.  the (ordered) subset of this belonging to `values`, which becomes `new_codes`, is:
```
[2, 3, 4, 6]
```
and the equivalent value for `target_values`, which becomes `new_target_codes`, is:
```
[0, 1, 3, 5, 7]
```

thus, if we are getting a backfilled indexer of `target_values` into `values`, we can simply backfill `new_target_codes` into `new_codes`, ge
tting us:
```
[0, 0, 1, 3, -1]
```
similarly, for padding, we can simply call `algos.pad(new_codes, new_target_codes)`, getting us:
```
[-1, -1, 1, 2, 3]
```
which can be seen to be correct

ChrisRobo commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,112 @@ cdef class BaseMultiIndexCodesEngine:
+        # the effect of this is to create a factorization for the (sorted)
+        # merger of the index values, where `new_codes` and `new_target_codes`
+        # are the subset of the factors which appear in `values` and `target`,
+        # respectively
```
example of how this code works:

if `values` is
```
[
    (1, 1),
    (1, 3),
    (2, 1),
    (2, 3),
]
```

and `target_values` (i.e. `target.values`, sorted) is:
```
[
   (0, 1),
   (1, 0),
   (1, 3),
   (2, 2),
   (2, 5),
]
```

then, if you merged these into a sorted list of (unique) values, you would get (including index values here, and an X indicates membership):
```
          values  target_values

0: (0, 1)               X
1: (1, 0)               X
2: (1, 1)    X
3: (1, 3)    X          X
4: (2, 1)    X
5: (2, 2)               X
6: (2, 3)    X
7: (2, 5)               X
```
and so we can think of this as a factorization of the merged and sorted list of tuples.  the (ordered) subset of this belonging to `values`, which becomes `new_codes`, is:
```
[2, 3, 4, 6]
```
and the equivalent value for `target_values`, which becomes `new_target_codes`, is:
```
[0, 1, 3, 5, 7]
```

thus, if we are getting a backfilled indexer of `target_values` into `values`, we can simply backfill `new_target_codes` into `new_codes`, ge
tting us:
```
[0, 0, 1, 3, -1]
```
similarly, for padding, we can simply call `algos.pad(new_codes, new_target_codes)`, getting us:
```
[-1, -1, 1, 2, 3]
```
which can be seen to be correct

ChrisRobo commented on pandas/core/indexes/multi.py:
```diff
@@ -2351,7 +2351,9 @@ def get_indexer(self, target, method=None, limit=None, tolerance=None):
                 )
-            indexer = self._engine.get_indexer(target, method, limit)
+            indexer = self._engine.__class__.get_indexer_and_fill(
+                self.values, target, method=method, limit=limit
```
I thought it would be preferable (per @jbrockmendel's point about keeping `Index` objects out of the cython code) to just pass the MultiIndex object's tuples directly to the cython code

ChrisRobo commented on pandas/tests/frame/indexing/test_indexing.py:
```diff
@@ -1400,6 +1400,79 @@ def test_set_value_resize(self, float_frame):
+        # 1:   2.0
+        # 2:   5.0
+        # 3:   5.8
+        df = pd.DataFrame({
```
same as before but added 3 0-level values for `df`'s (multi)index

ChrisRobo committed:
ensure type assignments correct

ChrisRobo committed:
use preferred method of getting class from object

ChrisRobo committed:
add newlines to get docs to compile and unindent for clarity

ChrisRobo committed:
add clarifying comment re: regression

ChrisRobo committed:
add black formatting

ChrisRobo committed:
clean up after slightly-botched rebase

ChrisRobo committed:
take suggestions of black pandas

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,111 @@ cdef class BaseMultiIndexCodesEngine:
+        target_order = np.argsort(target.values).astype('int64')
+        target_values = target.values[target_order]
+        num_values, num_target_values = len(values), len(target_values)
+        new_codes, new_target_codes = \
```
use parens rather than ``\``

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,111 @@ cdef class BaseMultiIndexCodesEngine:
-        else:
-            indexer = self._base.get_indexer(self, lab_ints)
+    @staticmethod
+    def get_indexer_and_fill(object values, object target,
```
don't use staticmethods

ChrisRobo replied:
rationale here was that it makes clear that `self` is not used; reverted anyway

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,111 @@ cdef class BaseMultiIndexCodesEngine:
-    def get_indexer(self, object target, object method=None,
-                    object limit=None):
+    def get_indexer(self, object target) -> np.ndarray:
+        """
```
why are you changing this api? can't you rather just call a helper function if you see method / limit?

ChrisRobo replied:
rationale is that in order to do reindexing with filling (not necessary when not filling though), without using "carrying" logic, it's necessary to know the full set of tuples, meaning we have a few options, as far as I can tell:

(1) pass the tuples in from the `MultiIndex` class.  The current implementation uses this; but I separated it for w/ and w/o filling to avoid adding new arguments to `get_indexer()`.

(2) since the MultiIndex's `_engine` knows only of the levels and the int representation of the codes, we could reconstruct the tuples from this, after doing appropriate place-value logic (i.e. inverting `self._codes_to_ints()`, see [1] for more detail if desired), and then proceed with the algorithm in `get_indexer_and_fill()`

(3) have the MultiIndex's engine store a reference to the tuples as well, but I thought a change like that might be inappropriate for this PR.  Happy to do whatever you prefer here!


**footnotes, not necessary to read**:
[1] this is basically the inversion of the `_codes_to_ints()` method in the `MultiIndex` class (source code [here](https://github.com/pandas-dev/pandas/blob/master/pandas/core/indexes/multi.py#L85-L113)).  I'd like to avoid doing this if possible, since it would be a (complex) computation of a value which can just be passed in, but it's definitely mathematically possible, and here's an example of how to do it, if this is unclear.

given:
```
>>> mi = pd.MultiIndex.from_tuples([
...     (1, 3),
...     (2, 1),
...     (2, 2),
...     (2, 4),
...     (5, 0)
... ])
>>> mi
MultiIndex([(1, 3),
            (2, 1),
            (2, 2),
            (2, 4),
            (5, 0)],
           )
```
we have the following info available to us:
```
>>> mi._engine.levels
FrozenList([[1, 2, 5], [0, 1, 2, 3, 4]])
>>> mi._engine.vgetter()
array([12, 18, 19, 21, 25], dtype=uint64)
```
knowing this, we can reconstruct that we need ceil(log_2(3)) = 2 bits for the 0-level values, ceil(log_2(5)) = 3 bits for the 1-level values, and so we could determine that:
- 12 = 1(2^3) + 4(2^0)
- 18 = 2(2^3) + 2(2^0)
- 19 = 2(2^3) + 3(2^0)
- 21 = 2(2^3) + 5(2^0)
- 25 = 3(2^3) + 1(2^0)

and so, after decrementing all of those coefficients by 1 to account for 0-indexing, this would give us the following (per-level) indexer-like list of indexes:
- (0, 3)
- (1, 1)
- (1, 2)
- (1, 4)
- (2, 0)
and if we get the values of `mi._engine.levels` at these (per-level) indices, we get:
- (1, 3)
- (2, 1)
- (2, 2)
- (2, 4)
- (5, 0)
i.e. the MultiIndex's list of tuples

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,111 @@ cdef class BaseMultiIndexCodesEngine:
+        np.ndarray[int64_t, ndim=1] of the indexer of `target` into `values`,
+        filled with the `method` (and optionally `limit`) specified
+        """
+        if method not in ("backfill", "pad"):
```
this should be in the calling method not here

jreback replied:
you can assert instead

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,111 @@ cdef class BaseMultiIndexCodesEngine:
+                new_target_codes[j] = next_code
+                j += 1
+            next_code += 1
+        # at this point, at least one should have reached the end
```
can you put a blank line before comments

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,111 @@ cdef class BaseMultiIndexCodesEngine:
+            next_code += 1
+
+        # get the indexer, and undo the sorting of `target.values`
+        sorted_indexer = \
```
same here use parens

jreback reviewed (changes requested):
some comments. pretty understadable. can you run the multi asv's and see if anything changes?

ChrisRobo committed:
address comments

ChrisRobo commented:
@jreback I ran the MultiIndex asvs (exact invocation was: `asv continuous -f 1.1 -E virtualenv upstream/master HEAD -b ^multiindex`) and the result was
```
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```
I can upload (or paste a subset of) the full multiindex benchmarks if you like as well

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,107 @@ cdef class BaseMultiIndexCodesEngine:
+        """
+        Gets an indexer, i.e. set of indexes into `self`'s values for the
+        values in `target`, where -1 represents a value in `target` not existing
+        in (the cross-product of) `self.levels`
```
This sentence is awkward.  How about something like "Return an array giving the positions of each value of `target` in `self.values`"?

ChrisRobo replied:
updated here and in the other docstring.  Let me know if you'd like further revision

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,107 @@ cdef class BaseMultiIndexCodesEngine:
+
+        Returns
+        -------
+        1-dimensional array of dtype int64 of the index
```
the pattern we use in the cython files would just be `ndarray[int64_t, ndim=1]`

ChrisRobo replied:
updated

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,107 @@ cdef class BaseMultiIndexCodesEngine:
+                             object method, object limit = None) -> np.ndarray:
+        """
+        Gets an indexer, i.e. a set of indexes into `values`, for the values in
+        `target`, where the index value.
```
same comment about akward sentence.

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,107 @@ cdef class BaseMultiIndexCodesEngine:
+        # are the subset of the factors which appear in `values` and `target`,
+        # respectively
+        i, j, next_code = 0, 0, 0
+        while i < num_values and j < num_target_values:
```
in python-space a while loop is slower than a for loop.  no idea if it matters in cython

ChrisRobo replied:
As far as I can tell, it's just a question of how it compiles.  I'd guess that a for-loop with `break` statements would compile similarly, and that the only difference here would be the location of the `jne` statements.  If you'd like, though, I can either profile it and/or send you a diff of the relevant `objdump` snippets?

jbrockmendel commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,107 @@ cdef class BaseMultiIndexCodesEngine:
+        # get the indexer, and undo the sorting of `target.values`
+        sorted_indexer = (
+            algos.backfill if method == "backfill" else algos.pad
+        )(new_codes, new_target_codes, limit=limit).astype('int64')
```
is there a compelling reason to do the algos call here rather than in the calling function?  theres no real perf benefit, and it looks like we could avoid needing method/limit kwargs here

ChrisRobo replied:
I did this to preserve the existing abstractions (i.e. MultiIndex defers to its engine on the implementation for `get_indexer()`), but if you'd prefer, I can definitely just change this to return the codes which will be the inputs to `algos.(pad|backfill)`

jbrockmendel replied:
not a strong preference

ChrisRobo committed:
update docstrings

ChrisRobo commented:
@jreback @jbrockmendel I think I've addressed all comments here, would one or both of you mind taking another look whenever you get a chance?

ChrisRobo commented:
ping on this @jreback @jbrockmendel -- does it look OK?  I think I've made all the desired updates, but happy to tweak more as is necessary.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -215,6 +215,67 @@ MultiIndex
 
+- Bug in :meth:`MultiIndex.get_indexer` incorrectly handling use of pad and backfill options.  This is a regression introduced in pandas 0.23.0 (:issue:`29896`)
+
+*pandas [0.23.0, 1.1.0)*
```
can you indicate that this is < 1.1.0 (the interval notation is too subtle here)

ChrisRobo replied:
done

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -215,6 +215,67 @@ MultiIndex
+
+*pandas 1.1.0*:
+
+.. code-block:: python
```
you can use an ipython-block to actually execute the code

ChrisRobo replied:
done

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -215,6 +215,67 @@ MultiIndex
+
+.. code-block:: python
+
+    >>> df = pd.DataFrame({
```
use an ipython block to execute & show the df (leave the reindex and so on);

use ipython prompts rather than ```>>>``` see other whatsnew sections

ChrisRobo replied:
a bit confused here -- presumably there's only one pandas environment in which it can execute this code?  It would make sense if that were the current version and so I could use that for the 1.1.0 block but I'm not sure how to programmatically get the docs to execute a code block on an older pandas version

jreback replied:
for the previous  section, we use static code, e.g. you execute it yourself in an older environment and just paste the output, the build *does* not execute it (hence the code-block), use ipython based formatting (see other entries in the whatsnew on how to do this).

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,110 @@ cdef class BaseMultiIndexCodesEngine:
 
-    def get_indexer(self, object target, object method=None,
-                    object limit=None):
+    def get_indexer(self, object target) -> np.ndarray:
```
I would rename this method, call it get_indexer_no_fill or something to differentiate from get_indexer (where we have elsewhere)

ChrisRobo replied:
done

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,110 @@ cdef class BaseMultiIndexCodesEngine:
-            indexer = indexer[order]
-        else:
-            indexer = self._base.get_indexer(self, lab_ints)
+    def get_indexer_and_fill(self, object values, object target,
```
rename this back to get_indexer

jreback replied:
and dispatch  if method is None (to get_indexer_no_fill), this keeps the api intact

ChrisRobo replied:
done, but needed to make the `values` arg optional (and used a different ordering in the invocation for when we are filling) to keep the api intact

jreback reviewed (changes requested):
some docs & api comments. looks good codewise. ping on green.

jreback commented on pandas/_libs/index.pyx:
```diff
@@ -606,25 +606,110 @@ cdef class BaseMultiIndexCodesEngine:
+        Parameters
+        ----------
+        values : list-like of tuples
+            must be sorted and all have the same length.  Should be the set of
```
do we have any asv's for these cases? if not can you add, and show perf on master and this branch

ChrisRobo replied:
I added some and pushed.

I'll sum up the differences which I saw on the PR's main page

ChrisRobo committed:
use ipython directive for example with bugfix

ChrisRobo committed:
update formatting

ChrisRobo committed:
put get_indexer() calls into MultiIndex engine behind single function

ChrisRobo committed:
add performance benchmark for MultiIndex get_indexer with filling

ChrisRobo committed:
fix MultiIndex reference

ChrisRobo commented:
@jreback per your request, I added some ASV benchmarks for these cases.

In summary, for the new benchmarks I added:
- this improves performance (> 10%) over 0.22.0, i.e. the last version where this feature worked
```
       before           after         ratio
     [a00154dc]       [0c3c0cbd]
     <v0.22.0^0>       <crobo/fix-multi-index-reindex>
...
-         552±6ms          477±3ms     0.86  multiindex_object.Integer.time_get_indexer_and_backfill
-         626±4ms          482±4ms     0.77  multiindex_object.Integer.time_get_indexer_and_pad
```
- performance with the current version of this PR is much worse than the (currently buggy) master (as expected, there were no significant changes on any of the existing ASVs)
```
       before           after         ratio
     [ed7adcd5]       [0c3c0cbd]
     <master>         <crobo/fix-multi-index-reindex>
+         124±2ms          481±6ms     3.87  multiindex_object.Integer.time_get_indexer_and_backfill
+         127±5ms          477±7ms     3.76  multiindex_object.Integer.time_get_indexer_and_pad
```
- performance with the pre-rewrite version of this PR (i.e. more complicated) code was not significantly different from master
master:
```
[ 75.00%] · For pandas commit ed7adcd5 <master> (round 2/2):
...
[ 90.00%] ··· multiindex_object.Integer.time_get_indexer_and_backfill      123±6ms
[ 91.25%] ··· multiindex_object.Integer.time_get_indexer_and_pad            121±3ms
```
this PR, pre-simplifying re-write:
```
[ 50.00%] · For pandas commit 6c8f50a1 <crobo/fix-multi-index-reindex~19> (round 2/2):
...
[ 65.00%] ··· multiindex_object.Integer.time_get_indexer_and_backfill      122±5ms
[ 66.25%] ··· multiindex_object.Integer.time_get_indexer_and_pad           124±7ms
```

conceptually it is much simpler, IMO w/o a doubt, to implement it like this.  It definitely is slower, though.  I can take a crack at simplifying my initial implementation and making the algorithm clearer but I defer to you re: the tradeoff of performance vs. simplicity here.

ChrisRobo committed:
take black pandas suggestions

ChrisRobo committed:
use black formatter on asv profiling tests

ChrisRobo commented:
@jreback build is green, pinging as requested.  see info in my previous comment re: ASV performance at various commits; happy to make adjustments as desired.

jreback commented:
@ChrisRobo is the current code reflecting the above comment `this PR, pre-simplifying re-write:` ?
e.g. where we are now

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -215,6 +215,50 @@ MultiIndex
+
+*pandas 1.1.0*:
+
+.. ipython:: python
```
for the new we actually execute it, so this is good here.

however,

show the setupcode (df, mi_2) in an ipython block at the top (as they are the same in both version)

ChrisRobo replied:
done, thanks

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -215,6 +215,50 @@ MultiIndex
+            'c': ['A', 'B', 'C', 'D'],
+        }).set_index(['a', 'b'])
+        mi_2 = pd.MultiIndex.from_product([[0], [-1, 0, 1, 3, 4, 5]])
+        df.reindex(mi_2, method='backfill')
```
either do in separate blocks the backfill and pad or put some comments, this will render as one long thing and hard to read

ChrisRobo replied:
i put pad and backfill each in their own sections, can attach rendered html if desired

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -215,6 +215,50 @@ MultiIndex
         # Common elements are now guaranteed to be ordered by the left side
         left.intersection(right, sort=False)
 
+- Bug in :meth:`MultiIndex.get_indexer` incorrectly handling use of pad and backfill options.  This is a regression introduced in pandas 0.23.0 (:issue:`29896`)
```
don't put this in bug fixes, this needs a separate section in api breaking changes), with a sub-header, see other examples there.

ChrisRobo replied:
done

ChrisRobo commented:
> @ChrisRobo is the current code reflecting the above comment `this PR, pre-simplifying re-write:` ?
> e.g. where we are now

the code in the PR right now is reflected by the second ASV snippet there.   The "pre-simplifying rewrite" commend referred to the PR as of [this review](https://github.com/pandas-dev/pandas/pull/30766#pullrequestreview-355635473), which was commit `6c8f50a1` in this branch.

jbrockmendel commented:
@ChrisRobo caffeine is still kicking in so bear with me as I make sure i understand the current status.  A previous draft of this PR maintained performance but with more complex code, while the current state of the PR is simpler but takes a performance hit?

ChrisRobo committed:
move whatsnew docs into their own sub-heading in api changes

ChrisRobo commented:
> @ChrisRobo caffeine is still kicking in so bear with me as I make sure i understand the current status. A previous draft of this PR maintained performance but with more complex code, while the current state of the PR is simpler but takes a performance hit?

hahah re: caffeine and yes precisely.  FWIW though, even the simpler but less performant code is a > 10% performance improvement over the last version (0.22.0) when this feature behaved correctly.

ChrisRobo committed:
fix syntax error

ChrisRobo committed:
rm duplicate test from failed rebase

ChrisRobo commented:
CI currently failing on a parquet test (see failure log [here](https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=31737&view=logs&j=a3a13ea8-7cf0-5bdb-71bb-6ac8830ae35c&t=add65f64-6c25-5783-8fd6-d9aa1b63d9d4)).  Not exactly sure how it's related to this change but I'll work on debugging

ChrisRobo commented:
ok, thanks to @jbrockmendel the CI issue seems to be fixed and the build is green.  Do you and @jreback have a sense of how you'd like to proceed here?  I've made all of @jreback's desired changes to the docs, I believe, but not sure how you want to proceed re: the performance/clarity questions

jreback commented on pandas/tests/frame/indexing/test_indexing.py:
```diff
@@ -1465,6 +1465,81 @@ def test_set_value_resize(self, float_frame):
         with pytest.raises(ValueError, match=msg):
             res._set_value("foobar", "baz", "sam")
 
+    def test_reindex_with_multi_index(self):
```
@jbrockmendel we should create a test_multi.py in this dir (@ChrisRobo you can do this if you want here)

jbrockmendel replied:
we're moving towards organizing by method instead of by index subclass, so this will eventually live in a test_reindex file (there are lots of reindex tests in this file, so this is fine for the time being.  before long ill batch-move them)

jreback reviewed (changes requested):
looks fine, can you merge master and just confirm the performance of what you have here. 

sorry this took a long time to review, we have *many* PRs. thanks

ChrisRobo committed:
clean up rebase error

ChrisRobo commented:
> looks fine, can you merge master and just confirm the performance of what you have here.
> 
> sorry this took a long time to review, we have _many_ PRs. thanks

rebased on top of master, perf for relevant ASV benchmarks seems the same as before

against master:
```
       before           after         ratio
     [e0dd81a5]       [7836cc90]
     <crobo/fix-multi-index-reindex~49>       <crobo/fix-multi-index-reindex>
+         120±7ms          484±9ms     4.05  multiindex_object.Integer.time_get_indexer_and_backfill
+         120±6ms          474±9ms     3.96  multiindex_object.Integer.time_get_indexer_and_pad
-        711±10μs          610±5μs     0.86  multiindex_object.Values.time_datetime_level_values_sliced
```

against 0.22.0 (last version where the correct functionality was in place):
```
       before           after         ratio
     [a00154dc]       [7836cc90]
     <v0.22.0^0>       <list>    
...
-        556±10ms          485±9ms     0.87  multiindex_object.Integer.time_get_indexer_and_backfill
-        626±10ms          482±5ms     0.77  multiindex_object.Integer.time_get_indexer_and_pad
```

and no problem re: reviews, thanks for taking the time to look at this!

jreback commented:
thanks @ChrisRobo great effort! and thanks for sticking with it. we are getting lots of PRs now-a-days and hard to give good review on them all.

ChrisRobo commented:
my pleasure, and thanks @jreback and @jbrockmendel for the time reviewing!
